File: src/renderer/gpu/digital.ts
import { GPUContext } from "./device";
import { Viewport, writeViewportInto, VIEWPORT_BYTES } from "./data";
import WGSL from "./digital.wgsl";

type ShaderVariant = "multi" | "single";

// RowInfo is 7×u32 (see segments.zig / digital.wgsl): x0_offset, x1_offset,
// bytes_per_sample, segment_start, flags, y_offset, height. Word 4 is the per-row
// flags (bit 0 = dim); words 5/6 are the vertical layout (CSS px as f32 bits),
// both patched directly by the renderer (no repack). ROW_FLAG_DIM must match
// digital.wgsl's ROW_FLAG_DIM.
const ROW_INFO_WORDS = 7;
const ROW_FLAG_DIM = 1 << 0;
const ROW_WORD_FLAGS = 4;
const ROW_WORD_Y = 5;
const ROW_WORD_H = 6;

export interface SignalPipeline {
  pipeline: GPURenderPipeline;
  bindGroup: GPUBindGroup;
  segmentCount: number;
  // The per-variant segment storage buffer backing this pipeline's bind group.
  // Tracked so a scene rebuild (add/remove active signal) can destroy it.
  segmentBuf: GPUBuffer;
}

// Shared per-scene buffers consumed by both single/multi pipelines.
export interface SceneBuffers {
  rowInfo: GPUBuffer;
  x0Pool: GPUBuffer;
  x1Pool: GPUBuffer;
  // CPU-side copy of the rowInfo records, retained so setDimFlags can patch the
  // per-row flags word and re-upload without a repack.
  rowInfoCpu: Uint32Array<ArrayBuffer>;
}

export interface DigitalRenderer {
  ctx: GPUContext;
  module: GPUShaderModule;
  bgl: GPUBindGroupLayout;
  layout: GPUPipelineLayout;
  uniformBuf: GPUBuffer;
  viewportScratch: Float32Array;
  writeViewport(vp: Viewport): void;
  createSceneBuffers(rowInfo: ArrayBuffer, x0Pool: ArrayBuffer, x1Pool: ArrayBuffer): SceneBuffers;
  buildPipelineFromPacked(
    variant: ShaderVariant,
    packed: Uint32Array<ArrayBuffer>,
    segmentCount: number,
    colorBuf: GPUBuffer,
    scene: SceneBuffers,
  ): Promise<SignalPipeline>;
  // Reuse an already-compiled pipeline with a fresh segment buffer + scene
  // (rebound bind group). Synchronous — no pipeline recompile — so an active
  // signal add/remove can repack on the spot. Caller owns destroying the old
  // SignalPipeline.segmentBuf and SceneBuffers.
  rebindPipeline(
    prev: SignalPipeline,
    packed: Uint32Array<ArrayBuffer>,
    segmentCount: number,
    colorBuf: GPUBuffer,
    scene: SceneBuffers,
  ): SignalPipeline;
  // Set the per-row dim flag (eye toggle) by patching the rowInfo buffer's flags
  // column and re-uploading it. One small writeBuffer, no repack — call after a
  // scene (re)build and whenever the hidden set changes.
  setDimFlags(scene: SceneBuffers, isHidden: (row: number) => boolean): void;
  // Write the per-row vertical layout (y_offset/height as f32 bits) into the
  // rowInfo buffer. `top` is the first row's y (below the ruler); heights stack.
  // `gapBelowOf` adds empty space after a row (divider) without growing its drawn
  // height. One writeBuffer, no repack — call after a scene (re)build / row resize.
  setRowLayout(scene: SceneBuffers, heightOf: (row: number) => number, top: number, gapBelowOf?: (row: number) => number): void;
}

export function createDigitalRenderer(ctx: GPUContext): DigitalRenderer {
  const { device, format } = ctx;

  const module = device.createShaderModule({ code: WGSL });

  const bgl = device.createBindGroupLayout({
    entries: [
      { binding: 0, visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT, buffer: { type: "uniform" } },
      { binding: 1, visibility: GPUShaderStage.VERTEX, buffer: { type: "read-only-storage" } },
      { binding: 2, visibility: GPUShaderStage.VERTEX, buffer: { type: "read-only-storage" } },
      { binding: 3, visibility: GPUShaderStage.VERTEX, buffer: { type: "read-only-storage" } },
      { binding: 4, visibility: GPUShaderStage.VERTEX, buffer: { type: "read-only-storage" } },
      { binding: 5, visibility: GPUShaderStage.VERTEX, buffer: { type: "read-only-storage" } },
    ],
  });

  const layout = device.createPipelineLayout({ bindGroupLayouts: [bgl] });

  const uniformBuf = device.createBuffer({ size: VIEWPORT_BYTES, usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST });
  const viewportScratch = new Float32Array(VIEWPORT_BYTES / 4);
  const viewportScratchI32 = new Int32Array(viewportScratch.buffer);

  const blend = {
    color: { srcFactor: "src-alpha", dstFactor: "one-minus-src-alpha", operation: "add" },
    alpha: { srcFactor: "one", dstFactor: "one-minus-src-alpha", operation: "add" },
  } as const;

  function writeStorage(bytes: ArrayBuffer, minSize = 16): GPUBuffer {
    // WebGPU validates each binding against the pipeline's required size even
    // for a 0-instance draw. An empty scene (fresh trace, nothing active) yields
    // an empty rowInfo buffer; binding 3 is array<RowInfo> (20 B stride), so it
    // must pad to one stride (minSize = 20) or validation rejects it as "too
    // small". The u32 pools use the default minSize 16 (harmless).
    const size = Math.max(minSize, bytes.byteLength);
    const buf = device.createBuffer({ size, usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST });
    if (bytes.byteLength > 0) device.queue.writeBuffer(buf, 0, bytes);
    return buf;
  }

  function createSceneBuffers(rowInfo: ArrayBuffer, x0Pool: ArrayBuffer, x1Pool: ArrayBuffer): SceneBuffers {
    return {
      rowInfo: writeStorage(rowInfo, ROW_INFO_WORDS * 4),
      x0Pool: writeStorage(x0Pool),
      x1Pool: writeStorage(x1Pool),
      rowInfoCpu: new Uint32Array(rowInfo),
    };
  }

  async function buildPipelineFromPacked(
    variant: ShaderVariant,
    packed: Uint32Array<ArrayBuffer>,
    segmentCount: number,
    colorBuf: GPUBuffer,
    scene: SceneBuffers,
  ): Promise<SignalPipeline> {
    const fragmentEntryPoint = variant === "single" ? "fs_single" : "fs_multi";
    const variantConst = variant === "single" ? 0 : 1;

    const pipeline = await device.createRenderPipelineAsync({
      layout,
      vertex: { module, entryPoint: "vs_main", constants: { VARIANT: variantConst } },
      fragment: {
        module,
        entryPoint: fragmentEntryPoint,
        constants: { VARIANT: variantConst },
        targets: [{ format, blend }],
      },
      primitive: { topology: "triangle-strip" },
    });

    return bindSegments(pipeline, packed, segmentCount, colorBuf, scene);
  }

  // Build the segment storage buffer + bind group for a (re)used pipeline.
  function bindSegments(
    pipeline: GPURenderPipeline,
    packed: Uint32Array<ArrayBuffer>,
    segmentCount: number,
    colorBuf: GPUBuffer,
    scene: SceneBuffers,
  ): SignalPipeline {
    const segmentBuf = device.createBuffer({
      size: Math.max(16, packed.byteLength),
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
    });
    if (packed.byteLength > 0) device.queue.writeBuffer(segmentBuf, 0, packed);

    const bindGroup = device.createBindGroup({
      label: "digital-bindgroup",
      layout: bgl,
      entries: [
        { binding: 0, resource: { buffer: uniformBuf } },
        { binding: 1, resource: { buffer: segmentBuf } },
        { binding: 2, resource: { buffer: colorBuf } },
        { binding: 3, resource: { buffer: scene.rowInfo } },
        { binding: 4, resource: { buffer: scene.x0Pool } },
        { binding: 5, resource: { buffer: scene.x1Pool } },
      ],
    });

    return { pipeline, bindGroup, segmentCount, segmentBuf };
  }

  function rebindPipeline(
    prev: SignalPipeline,
    packed: Uint32Array<ArrayBuffer>,
    segmentCount: number,
    colorBuf: GPUBuffer,
    scene: SceneBuffers,
  ): SignalPipeline {
    return bindSegments(prev.pipeline, packed, segmentCount, colorBuf, scene);
  }

  function writeViewport(vp: Viewport): void {
    writeViewportInto(viewportScratch, viewportScratchI32, vp);
    device.queue.writeBuffer(uniformBuf, 0, viewportScratch);
  }

  function setDimFlags(scene: SceneBuffers, isHidden: (row: number) => boolean): void {
    const cpu = scene.rowInfoCpu;
    const rows = cpu.length / ROW_INFO_WORDS;
    for (let r = 0; r < rows; r++) {
      cpu[r * ROW_INFO_WORDS + ROW_WORD_FLAGS] = isHidden(r) ? ROW_FLAG_DIM : 0;
    }
    if (cpu.byteLength > 0) device.queue.writeBuffer(scene.rowInfo, 0, cpu);
  }

  function setRowLayout(scene: SceneBuffers, heightOf: (row: number) => number, top: number, gapBelowOf?: (row: number) => number): void {
    const cpu = scene.rowInfoCpu;
    const rows = cpu.length / ROW_INFO_WORDS;
    // Same backing buffer, viewed as f32 to write the y_offset/height words.
    const f = new Float32Array(cpu.buffer, cpu.byteOffset, cpu.length);
    let y = top;
    for (let r = 0; r < rows; r++) {
      const h = heightOf(r);
      f[r * ROW_INFO_WORDS + ROW_WORD_Y] = y;
      f[r * ROW_INFO_WORDS + ROW_WORD_H] = h;
      y += h + (gapBelowOf ? gapBelowOf(r) : 0);
    }
    if (cpu.byteLength > 0) device.queue.writeBuffer(scene.rowInfo, 0, cpu);
  }

  return { ctx, module, bgl, layout, uniformBuf, viewportScratch, writeViewport, createSceneBuffers, buildPipelineFromPacked, rebindPipeline, setDimFlags, setRowLayout };
}

File: crates/riptide-render/src/digital.rs
//! The digital waveform pipelines (port of `gpu/digital.ts` + `digital.wgsl`):
//! one shader module, a `VARIANT` override picks single (1-bit line) vs multi
//! (pill) at pipeline build; triangle-strip rects, 4 verts × N instances;
//! 6 bind-group entries (viewport uniform + segments/colors/rows/x0/x1
//! read-only storage); alpha blending on.
//!
//! OWNED BY UNIT U6 — which also owns the migration's single semantic WGSL
//! edit: the `F_HATCH_COLOR` predicate flip in `shaders/digital.wgsl` for the
//! tide.rs X/Z plane swap (see MIGRATION.md).

use riptide_contract::gpu::PackedSegment;

use crate::colors::ColorBuffer;
use crate::device::{Gpu, ViewportBuffer};
use crate::scene::SceneBuffers;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Variant {
    Single,
    Multi,
}

/// Shader module + layouts + the two compiled pipelines, built once per
/// device/format. Both variants are compiled eagerly in [`DigitalContext::new`]
/// so [`DigitalContext::bind`] is synchronous (the old `rebindPipeline` fast
/// path is the only path).
#[non_exhaustive]
pub struct DigitalContext {
    pub module: wgpu::ShaderModule,
    pub bgl: wgpu::BindGroupLayout,
    pub layout: wgpu::PipelineLayout,
    single: wgpu::RenderPipeline,
    multi: wgpu::RenderPipeline,
}

/// One variant's bound draw state: segment storage buffer + bind group.
/// Rebuilt (rebind, no recompile) on every repack.
#[non_exhaustive]
pub struct SignalPipeline {
    pipeline: wgpu::RenderPipeline,
    bind_group: wgpu::BindGroup,
    /// The per-variant segment storage buffer backing this pipeline's bind
    /// group (the label batches bind it too).
    pub segment_buf: wgpu::Buffer,
    pub segment_count: u32,
}

/// Alpha blend state from `digital.ts`: straight src-alpha over for color,
/// (one, one-minus-src-alpha) for alpha.
const BLEND: wgpu::BlendState = wgpu::BlendState {
    color: wgpu::BlendComponent {
        src_factor: wgpu::BlendFactor::SrcAlpha,
        dst_factor: wgpu::BlendFactor::OneMinusSrcAlpha,
        operation: wgpu::BlendOperation::Add,
    },
    alpha: wgpu::BlendComponent {
        src_factor: wgpu::BlendFactor::One,
        dst_factor: wgpu::BlendFactor::OneMinusSrcAlpha,
        operation: wgpu::BlendOperation::Add,
    },
};

fn build_pipeline(
    gpu: &Gpu,
    module: &wgpu::ShaderModule,
    layout: &wgpu::PipelineLayout,
    variant: Variant,
) -> wgpu::RenderPipeline {
    let (label, fs_entry, variant_const) = match variant {
        Variant::Single => ("digital-single", "fs_single", 0.0),
        Variant::Multi => ("digital-multi", "fs_multi", 1.0),
    };
    let constants: &[(&str, f64)] = &[("VARIANT", variant_const)];
    gpu.device
        .create_render_pipeline(&wgpu::RenderPipelineDescriptor {
            label: Some(label),
            layout: Some(layout),
            vertex: wgpu::VertexState {
                module,
                entry_point: Some("vs_main"),
                compilation_options: wgpu::PipelineCompilationOptions {
                    constants,
                    ..Default::default()
                },
                buffers: &[],
            },
            fragment: Some(wgpu::FragmentState {
                module,
                entry_point: Some(fs_entry),
                compilation_options: wgpu::PipelineCompilationOptions {
                    constants,
                    ..Default::default()
                },
                targets: &[Some(wgpu::ColorTargetState {
                    format: gpu.format,
                    blend: Some(BLEND),
                    write_mask: wgpu::ColorWrites::ALL,
                })],
            }),
            primitive: wgpu::PrimitiveState {
                topology: wgpu::PrimitiveTopology::TriangleStrip,
                ..Default::default()
            },
            depth_stencil: None,
            multisample: wgpu::MultisampleState::default(),
            multiview: None,
            cache: None,
        })
}

impl DigitalContext {
    pub fn new(gpu: &Gpu) -> Self {
        let module = gpu
            .device
            .create_shader_module(wgpu::ShaderModuleDescriptor {
                label: Some("digital-wgsl"),
                source: wgpu::ShaderSource::Wgsl(crate::DIGITAL_WGSL.into()),
            });

        // 6 bindings: viewport uniform (vs+fs) + segments/colors/rows/x0/x1
        // read-only storage (vs only) — mirrors digital.ts's bind group layout.
        let storage_entry = |binding: u32| wgpu::BindGroupLayoutEntry {
            binding,
            visibility: wgpu::ShaderStages::VERTEX,
            ty: wgpu::BindingType::Buffer {
                ty: wgpu::BufferBindingType::Storage { read_only: true },
                has_dynamic_offset: false,
                min_binding_size: None,
            },
            count: None,
        };
        let bgl = gpu
            .device
            .create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
                label: Some("digital-bgl"),
                entries: &[
                    wgpu::BindGroupLayoutEntry {
                        binding: 0,
                        visibility: wgpu::ShaderStages::VERTEX | wgpu::ShaderStages::FRAGMENT,
                        ty: wgpu::BindingType::Buffer {
                            ty: wgpu::BufferBindingType::Uniform,
                            has_dynamic_offset: false,
                            min_binding_size: None,
                        },
                        count: None,
                    },
                    storage_entry(1),
                    storage_entry(2),
                    storage_entry(3),
                    storage_entry(4),
                    storage_entry(5),
                ],
            });

        let layout = gpu
            .device
            .create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
                label: Some("digital-layout"),
                bind_group_layouts: &[&bgl],
                push_constant_ranges: &[],
            });

        let single = build_pipeline(gpu, &module, &layout, Variant::Single);
        let multi = build_pipeline(gpu, &module, &layout, Variant::Multi);

        Self {
            module,
            bgl,
            layout,
            single,
            multi,
        }
    }

    /// Builds the segment buffer + bind group for `segments` against the
    /// scene/viewport/colors buffers. Synchronous (pipelines pre-compiled in
    /// `new`) so add/remove repacks rebind on the spot — the old
    /// `rebindPipeline` fast path is the only path.
    pub fn bind(
        &self,
        gpu: &Gpu,
        variant: Variant,
        segments: &[PackedSegment],
        viewport: &ViewportBuffer,
        colors: &ColorBuffer,
        scene: &SceneBuffers,
    ) -> SignalPipeline {
        let bytes: &[u8] = bytemuck::cast_slice(segments);
        // An empty scene still validates: WebGPU/wgpu checks each binding
        // against the pipeline's required size even for a 0-instance draw, so
        // pad the (possibly empty) segment buffer to at least one stride
        // (Segment = 12 B; 16 keeps the digital.ts minimum).
        let size = (bytes.len() as u64).max(16);
        let segment_buf = gpu.device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("digital-segments"),
            size,
            usage: wgpu::BufferUsages::STORAGE | wgpu::BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });
        if !bytes.is_empty() {
            gpu.queue.write_buffer(&segment_buf, 0, bytes);
        }

        let bind_group = gpu.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("digital-bindgroup"),
            layout: &self.bgl,
            entries: &[
                wgpu::BindGroupEntry {
                    binding: 0,
                    resource: viewport.buf.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 1,
                    resource: segment_buf.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 2,
                    resource: colors.buf.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 3,
                    resource: scene.row_info.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 4,
                    resource: scene.x0_pool.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 5,
                    resource: scene.x1_pool.as_entire_binding(),
                },
            ],
        });

        let pipeline = match variant {
            Variant::Single => self.single.clone(),
            Variant::Multi => self.multi.clone(),
        };

        SignalPipeline {
            pipeline,
            bind_group,
            segment_buf,
            segment_count: segments.len() as u32,
        }
    }
}

impl SignalPipeline {
    pub fn draw(&self, pass: &mut wgpu::RenderPass<'_>) {
        pass.set_pipeline(&self.pipeline);
        pass.set_bind_group(0, &self.bind_group, &[]);
        pass.draw(0..4, 0..self.segment_count);
    }
}
